src/wssession/session_manager.rs

```diff
@@ -0,0 +1,106 @@
+///
+/// ```ignore
+/// use tradier::wssession::session_manager::SessionManager;
+///
+/// let manager = SessionManager::default();
+/// assert!(manager.acquire_session().is_ok());
+/// assert!(manager.acquire_session().is_err()); // Second acquisition fails
+/// manager.release_session(); // Releases the session
+/// assert!(manager.acquire_session().is_ok()); // Session can be reacquired
+/// ```
+#[derive(Default, Debug)]
+pub(crate) struct SessionManager {
+    session_exists: AtomicBool,
+}
+
+impl SessionManager {
+    /// Attempts to acquire a session, ensuring only one active session at a time.
+    ///
+    /// This method atomically sets the session state to "active". If a session is already
+    /// active, it returns an error indicating that the session cannot be acquired.
+    ///
+    /// This method is for internal use by higher-level abstractions like [`Session`].
+    ///
+    /// # Returns
+    /// - `Ok(())` if the session was successfully acquired.
+    /// - `Err(Error::SessionAlreadyExists)` if a session is already active.
+    pub(crate) fn acquire_session(&self) -> Result<()> {
+        if self
+            .session_exists
+            .compare_exchange(false, true, Ordering::Acquire, Ordering::Relaxed)
+            .is_err()
+        {
+            return Err(Error::SessionAlreadyExists);
+        }
+        Ok(())
+    }
+
+    /// Releases the active session, allowing a new session to be acquired.
+    ///
+    /// This method atomically sets the session state to "inactive". It is for internal
+    /// use by components like [`Session`] to manage session lifecycles.
+    pub(crate) fn release_session(&self) {
+        self.session_exists.store(false, Ordering::Release);
+    }
+}
+
+impl Drop for SessionManager {
+    /// Automatically releases the session when the `SessionManager` is dropped.
+    ///
+    /// This ensures that any active session is released when a `SessionManager` instance
+    /// goes out of scope. This method is primarily for internal use.
+    fn drop(&mut self) {
+        self.release_session();
+    }
+}
+
+/// A globally accessible, lazily-initialized singleton instance of `SessionManager`.
+///
+/// The `GLOBAL_SESSION_MANAGER` provides a single, thread-safe instance of `SessionManager`
+/// for use within the crate. It is not directly accessible outside the crate due to its
+/// `pub(crate)` visibility.
+///
+/// Most session-related operations should use higher-level abstractions like [`Session`].
+///
+/// ### Key Features
+/// - **Lazy Initialization**: The `SessionManager` is created only when first accessed.
+/// - **Thread-Safe**: Ensures safe, concurrent access across threads.
+///
+/// ### Internal Usage
+///
+/// Direct access to `GLOBAL_SESSION_MANAGER` is intended only for crate-internal use. External
+/// consumers of the library should use public abstractions like [`Session`].
+pub(crate) static GLOBAL_SESSION_MANAGER: LazyLock<SessionManager> = LazyLock::new(|| {
```

Comment:
The current implementation creates an unnecessary closure. Since `SessionManager::default` is already a function that matches the required signature, we can pass it directly to `LazyLock::new`. This:

- Makes the code more concise and idiomatic
- Avoids creating a redundant closure wrapper
- Leverages Rust's first-class functions
- Is slightly more efficient (avoids closure allocation)

Consider changing:
```rust
pub(crate) static GLOBAL_SESSION_MANAGER: LazyLock<SessionManager> =
    LazyLock::new(SessionManager::default);
```

Reply:
I am totally going to do this. But, from these docs I am not seeing a call out about the slight performance hit of an unnecessary closure: https://doc.rust-lang.org/book/ch19-05-advanced-functions-and-closures.html

Am I not looking in the right place?

Reply:
lint is complaining about that as well:

```
error: redundant closure
   --> src/wssession/session_manager.rs:105:19
    |
105 |     LazyLock::new(|| SessionManager::default());
    |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace the closure with the function itself: `SessionManager::default`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
    = note: `-D clippy::redundant-closure` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::redundant_closure)]`

error: could not compile `tradier` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
make: *** [Makefile:34: lint] Error 101
Error: Process completed with exit code 2.
```